src/Processors/Transforms/DeduplicationTokenTransforms.cpp

```diff
@@ -0,0 +1,241 @@
+#include <Processors/Transforms/DeduplicationTokenTransforms.h>
+
+#include <IO/WriteHelpers.h>
+
+#include <Common/logger_useful.h>
+#include <Common/Exception.h>
+#include <Common/SipHash.h>
+
+
+#include <fmt/core.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+void RestoreChunkInfosTransform::transform(Chunk & chunk)
+{
+    chunk.getChunkInfos().append(chunk_infos.clone());
+}
+
+namespace DeduplicationToken
+{
+
+String TokenInfo::getToken() const
+{
+    if (!isDefined())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "token is not defined, stage {}, token {}", stage, debugToken());
+
+    return getTokenImpl();
+}
+
+String TokenInfo::getTokenImpl() const
+{
+    String result;
+    result.reserve(getTotalSize());
+
+    for (const auto & part : parts)
+    {
+        if (!result.empty())
+            result.append(":");
+        result.append(part);
+    }
+
+    return result;
+}
+
+String TokenInfo::debugToken() const
+{
+    return getTokenImpl();
+}
+
+void TokenInfo::addChunkHash(String part)
+{
+    if (stage == UNDEFINED)
+        stage = DEFINE_SOURCE_WITH_HASHES;
```

Comment:
Should we `assert(empty())` here if `stage == UNDEFINED`?

Reply:
changed. Txnx.